app/controllers/idv/session_errors_controller.rb

```diff
@@ -0,0 +1,45 @@
+module Idv
+  class SessionErrorsController < ApplicationController
+    include IdvSession
+
+    before_action :confirm_two_factor_authenticated_or_recovery
+    before_action :confirm_idv_session_step_needed
+
+    def warning
+      @remaining_step_attempts = remaining_step_attempts
+    end
+
+    def timeout
+      @remaining_step_attempts = remaining_step_attempts
+    end
+
+    def jobfail
+      @remaining_step_attempts = remaining_step_attempts
+    end
+
+    def failure; end
+
+    private
+
+    def confirm_two_factor_authenticated_or_recovery
+      user_signed_in? || session[:ial2_recovery_user_id].present?
```

Comment:
`user_signed_in?` needs to be `user_fully_authenticated?` or similar for a 2fa-ed and setup user.

Reply:
Also, this should def be covered by a test

Reply:
Cleaned this up